### PR TITLE
add user usage detail in nodes

### PR DESCRIPTION
**Pull request by zhongfq**
1. add usage detail icon in left-bottom
IMG_0650

2. toggle show usage in nodes.
IMG_0652

3. node usages(in menu list)
IMG_4DF7188E2623-1

### CODE REVIEW

**Comment by s4I3h** [comment]
@zhongfq

**Comment by M03ED** [comment]
wow , you make my idea and did it really fast , i didn't you could do this at short time.
thanks man.

**Comment by SaintShit** [comment]
it's working like clockwork! great job on the changes.

File: app/views/user.py [hunk 1]
```diff
@@ -209,3 +209,21 @@ def reset_users_data_usage(db: Session = Depends(get_db),
...
+             db: Session = Depends(get_db),
+             admin: Admin = Depends(Admin.get_current)):
+    """
+    Get users usage
+    """
+    dbuser = crud.get_user(db, username)
+    if not dbuser:
+        raise HTTPException(status_code=404, detail="User not found")
+
+    if not (admin.is_sudo or (dbuser.admin and dbuser.admin.username == admin.username)):
```
Comment (SaintShit) [inline app/views/user.py:224]: I'm not sure why that check is necessary. could you give me some information why this is important to be a sudo here?

Comment (zhongfq) [reply app/views/user.py:224]: 😂😂，it my faults, I copy the code of get_user. so it may be not need here

**Comment by zhongfq** [comment]
> wow , you make my idea and did it really fast , i didn't you could do this at short time.
> thanks man.

this feature is not difficult to implement

**Comment by meaqese** [comment]
> > wow , you make my idea and did it really fast , i didn't you could do this at short time.
> > thanks man.
> 
> this feature is not difficult to implement

+2065 lines....

**Comment by Alirezaja1384** [comment]
@zhongfq این چینیا روزاشون چند ساعته؟ هرجور حساب میکنم این یارو خیلی خفنه

**Comment by SaintShit** [comment]
Hello again, I'm going to make some changes in user_usages table
I want to change it's name to `node_user_usages`
change the user_id to user_username to be more straight for the recording job. and some other minor changes. just wanted to let u know, u might run `alembic downgrade -1` once and then upgrade again

**Comment by M03ED** [comment]
Can you add period choose ? 
Past 24H 
7 days 
30 days 
And numbers like this

**Comment by zhongfq** [comment]
> Hello again, I'm going to make some changes in user_usages table
> I want to change it's name to `node_user_usages`
> change the user_id to user_username to be more straight for the recording job. and some other minor changes. just wanted to let u know, u might run `alembic downgrade -1` once and then upgrade again

OK

**Comment by zhongfq** [comment]
@SaintShit because the master and slave node are very similar, can let master be used as  node 0(nodeid=0) to simplify statistics?

**Comment by zhongfq** [comment]
> Can you add period choose ?
> Past 24H
> 7 days
> 30 days
> And numbers like this

Finish the current pull request first, and I will try implement it later.

**Comment by ArshaShk** [comment]
WOW Great Job!
I'll be very appreciate if you also add usage multiplier for nodes
Like if you use 1gb from Germany node get 1gb from users traffic but if you use 1gb from United States node use 3gb from users traffic
It's good for different data centers like in hetzner you get 20tb traffic with 4$ but in digitalocean you get only 500gb

**Comment by SaintShit** [comment]
I was thinking about the best way to upsert to NodeUserUsages table, looks like we have no way except checking the not existing rows before the update query. I'm going to merge the current changes and then debug these.

**Comment by zhongfq** [comment]
> I was thinking about the best way to upsert to NodeUserUsages table, looks like we have no way except checking the not existing rows before the update query. I'm going to merge the current changes and then debug these.

I think we can cache the key(username + "|" + node_id)，it check only once

**Comment by zhongfq** [comment]
> > @SaintShit because the master and slave node are very similar, can let master be used as node 0(nodeid=0) to simplify statistics?
> 
> I think the way you implemented this is OK I'm running into some bugs and trying to figure it out why this is happening
> 
> image like this, after reseting users' usage, this happens

I test is now

**Comment by SaintShit** [comment]
> > I was thinking about the best way to upsert to NodeUserUsages table, looks like we have no way except checking the not existing rows before the update query. I'm going to merge the current changes and then debug these.
> 
> I think we can cache the key(username + "|" + node_id)，it check only once

Hmmm right, let's see if the current changes works well in large scale or not, if not, we'll add Redis. it will be used in other parts of code too ;D

**Comment by zhongfq** [comment]
> > @SaintShit because the master and slave node are very similar, can let master be used as node 0(nodeid=0) to simplify statistics?
> 
> I think the way you implemented this is OK I'm running into some bugs and trying to figure it out why this is happening
> 
> image like this, after reseting users' usage, this happens

I do check in my compute, it is ok. func 'reset_all_users_data_usage' is not clear the nodes and system stat, so I think it will not affect the nodes usages

**Comment by SaintShit** [comment]
> > > @SaintShit because the master and slave node are very similar, can let master be used as node 0(nodeid=0) to simplify statistics?
> > 
> > 
> > I think the way you implemented this is OK I'm running into some bugs and trying to figure it out why this is happening
> > image like this, after reseting users' usage, this happens
> 
> I do check in my compute, it is ok. func 'reset_all_users_data_usage' is not clear the nodes and system stat, so I think it will not affect the nodes usages

image
image

**Comment by SaintShit** [comment]
Yes, used_traffic has been reset, but node usages has not been reset. I'll check this now

**Comment by SaintShit** [comment]
> > image image
> 
> master = user - germany
> 
> so may be user stat is not correct

Now node_usages will reset too
15964ce69380bf527e8c7acdf69fd0b629a197a4

**Comment by SaintShit** [comment]
> > Now node_usages will reset too [15964ce](https://github.com/Gozargah/Marzban/commit/15964ce69380bf527e8c7acdf69fd0b629a197a4)
> 
> I have a question is, why when cleaning all users uages , do we not clean system usage and nodes usage in node table ?

Actually they're not the total usage of users, they are the amount of data that has passed through the outbounds. why should we reset them?

**Comment by zhongfq** [comment]
> Actually they're not the total usage of users, they are the amount of data that has passed through the outbounds. why should we reset them?

for example, hire vps in linode or other cloud sever, there have usage limitation in one month, usually reset at next month, so I can see how usage used in current period.

**Comment by SaintShit** [comment]
> 

so, we can add a feature to reset system usage, or a reset option for each node. is this gonna solve this?

**Comment by zhongfq** [comment]
> > Actually they're not the total usage of users, they are the amount of data that has passed through the outbounds. why should we reset them?
> > for example, hire vps in linode or other cloud sever, there have usage limitation in one month, usually reset at next month, so I can see how usage used in current period.
> 
> so, we can add a feature to reset system usage, or a reset option for each node. is this gonna solve this?

yes, I think we can add this feature in node setting panel.

also, we can add two another feature:
1. upgrade marzban-node:  git pull && systemctl restart marzban
2. upgrade xray with selected version: Xray-install.sh install version

**Comment by SaintShit** [comment]
> > > Actually they're not the total usage of users, they are the amount of data that has passed through the outbounds. why should we reset them?
> > > for example, hire vps in linode or other cloud sever, there have usage limitation in one month, usually reset at next month, so I can see how usage used in current period.
> > 
> > 
> > so, we can add a feature to reset system usage, or a reset option for each node. is this gonna solve this?
> 
> yes, I think we can add this feature in node setting panel.
> 
> also, we can add two another feature:
> 
> 1. upgrade marzban-node:  git pull && systemctl restart marzban
> 2. upgrade xray with selected version: Xray-install.sh install version

Yay, just made [this issue](https://github.com/Gozargah/Marzban-node/issues/4). 
the majority of our users use docker, as we suggested

**Pull request merged**